### PR TITLE
Spec versions: allow `git.` references for branches with `/`

### PR DESCRIPTION
**Pull request by scheibelp**
For example, you can ask for `spack spec foo@git.feature/x`, assuming that the Git repository for `foo` has a branch called `feature/x`. Without this PR, the `/` in `feature/x` gets parsed incorrectly.

### CODE REVIEW

File: lib/spack/spack/test/conftest.py [hunk 1]
```diff
@@ -145,70 +145,84 @@ def commit(message):
...
-        git("init")
+    def _make(branch_name=None):
+        branch_name = branch_name or "1.x"
+        with working_dir(repo_path):
+            git("init")
 
-        git("config", "user.name", "Spack")
-        git("config", "user.email", "spack@spack.io")
+            git("config", "user.name", "Spack")
+            git("config", "user.email", "spack@spack.io")
```
Comment (alalazo) [inline lib/spack/spack/test/conftest.py]: I wouldn't change this to make a mock git repo with a feature branch. This fixture is brittle, and requires external tools like `git` so I'd avoid it if possible in unit tests. What I'd do instead is to add further tests in `test/spec_syntax.py` and check that a given string literal is tokenized as we expect. 

Some examples of tests like that are:
https://github.com/spack/spack/blob/3a565c66e9fc0bd5050faf7d07834e20a23b458f/lib/spack/spack/test/spec_syntax.py#L409-L419

For instance, I see that on `develop` the literal `zlib@git.foo/bar` is incorrectly parsed as:
```python
>>> for token in spack.parser.tokenize("zlib@git.foo/bar"):
...     print(token)
... 
(UNQUALIFIED_PACKAGE_NAME, zlib)
(VERSION, @git.foo)
(DAG_HASH, /bar)
```
while in this branch is correctly:
```python
>>> for token in spack.parser.tokenize("zlib@git.foo/bar"):
...     print(token)
... 
(UNQUALIFIED_PACKAGE_NAME, zlib)
(GIT_VERSION, @git.foo/bar)
```

**Review by alalazo — CHANGES_REQUESTED** [review]
The fix LGTM. I have a comment on the unit test.

File: lib/spack/spack/test/conftest.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/test/conftest.py]: I agree that this is a good targeted test.

I felt that other behaviors like lookup etc. should also be tested, and this covered that. I scaled it back to a smaller test: `test_git_branch_with_slash`. Let me know if that seems sufficient and focused.

**Comment by scheibelp** [comment]
@alalazo I think I've addressed the review: could you review again?

File: lib/spack/spack/test/conftest.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/test/conftest.py]: Thanks @scheibelp !

**Review by alalazo — APPROVED** [review]

**Comment by haampie** [comment]
I'm surprised to see this PR get merged if it just adds parsing without a thought about the consequences.

`/` was not allowed because of file system projections. It's relied on in tons of places: database, buildcache, maybe modules, makefile generation...

I'd rather see this reverted and ensure that spec -> path / url mappings are safe.

For projections we'd probably need to split on the dir separator before interpolation, then interpolate each component, map unsafe characters to say `_`, and join again.

Apart from that you'd have to go over how urls for buildcaches etc are constructed computed, and maybe check if module files do anything custom, as well as views.

**Comment by alalazo** [comment]
I overlooked that `{version}` would result in multiple nested levels per directory. It seems to "work" for installation trees, though I agree it might not desirable (check `zlib-ng` below):

Screenshot from 2023-08-09 11-15-21

For what is worth we are looking into it with @haampie

**Comment by alalazo** [comment]
Buildcaches with this kind of git reference are failing to construct:
Screenshot from 2023-08-09 11-36-20

**Comment by haampie** [comment]
By just grepping I've found the following potentially problematic uses of `{version}` interpolation:

https://github.com/spack/spack/blob/a7f2abf9242ad321b410cc2f43b394856facba23/lib/spack/spack/filesystem_view.py#L503

https://github.com/spack/spack/blob/ba1d29502388ba85d9c45df4ac70c8b980a8720d/lib/spack/spack/stage.py#L52-L55

https://github.com/spack/spack/blob/95847a0b373d1fe55772ff39d58aa616bec88932/lib/spack/spack/build_systems/oneapi.py#L59-L62

https://github.com/spack/spack/blob/0b4631a7741df403510281b304d7bd6b169beebc/lib/spack/spack/cmd/ci.py#L580

https://github.com/spack/spack/blob/ba1d29502388ba85d9c45df4ac70c8b980a8720d/lib/spack/spack/package_base.py#L991-L993

https://github.com/spack/spack/blob/ba1d29502388ba85d9c45df4ac70c8b980a8720d/lib/spack/spack/package_base.py#L1003-L1005

Use of `spec.version`:

https://github.com/spack/spack/blob/37be51b9ddccf84a793ed6a6a7e6ba94dcd2ea59/lib/spack/spack/binary_distribution.py#L793-L813

https://github.com/spack/spack/blob/a7f2abf9242ad321b410cc2f43b394856facba23/lib/spack/spack/cmd/install.py#L243

https://github.com/spack/spack/blob/9aa3b4619b867e195ba74b0da6ed1db0039c331d/lib/spack/spack/modules/lmod.py#L287

---

Comment below is copied from #39354:

> All current build caches rely on a flat dir structure for reindexing, so the "accidental" subdir created by `git.x/y` version is problematic.
> 
> Suggestion is to fix file system projections roughly like:
> 
> ```python
> path_component_format = lambda component: re.sub(r"[^\w.-]", "_", spec.format(str))
> '/'.join(map(path_component_format(component, projection.split('/'))))
> ```
> 
> That also maps error-prone characters like  `' '` and `'='` to `'_'`, which I think is a good idea since we sometimes put `spec.prefix` without escaping into makefiles in the `def edit` phase.
> 
> On top of that it might be worth to disallow literal `..` components :upside_down_face:

**Pull request merged**